src/Workspaces/Core/Portable/Rename/IRemoteRenamer.cs

```diff
@@ -0,0 +1,208 @@
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    internal interface IRemoteRenamer
+    {
+        Task<SerializableRenameLocations> FindRenameLocationsAsync(
+            PinnedSolutionInfo solutionInfo, SerializableSymbolAndProjectId symbol, SerializableRenameOptionSet options, CancellationToken cancellationToken);
+    }
+
+    internal struct SerializableRenameOptionSet
+    {
+        public bool RenameOverloads;
+        public bool RenameInStrings;
+        public bool RenameInComments;
+        public bool RenameFile;
+
+        public static SerializableRenameOptionSet Dehydrate(RenameOptionSet optionSet)
+            => new SerializableRenameOptionSet
+            {
+                RenameOverloads = optionSet.RenameOverloads,
+                RenameInStrings = optionSet.RenameInStrings,
+                RenameInComments = optionSet.RenameInComments,
+                RenameFile = optionSet.RenameFile,
+            };
+
+        public RenameOptionSet Rehydrate()
+            => new RenameOptionSet(RenameOverloads, RenameInStrings, RenameInComments, RenameFile);
+    }
+
+    internal class SerializableSearchResult
+    {
+        // We use arrays so we can represent default immutable arrays.
+
+        public SerializableRenameLocation[] Locations;
+        public SerializableReferenceLocation[] ImplicitLocations;
+        public SerializableSymbolAndProjectId[] ReferencedSymbols;
+
+        public static SerializableSearchResult Dehydrate(Solution solution, RenameLocations.SearchResult result, CancellationToken cancellationToken)
+            => result == null ? null : new SerializableSearchResult
+            {
+                Locations = result.Locations?.Select(loc => SerializableRenameLocation.Dehydrate(loc)).ToArray(),
+                ImplicitLocations = result.ImplicitLocations.IsDefault ? null : result.ImplicitLocations.Select(loc => SerializableReferenceLocation.Dehydrate(loc, cancellationToken)).ToArray(),
+                ReferencedSymbols = result.ReferencedSymbols.IsDefault ? null : result.ReferencedSymbols.Select(s => SerializableSymbolAndProjectId.Dehydrate(solution, s, cancellationToken)).ToArray(),
+            };
+
+        public async Task<RenameLocations.SearchResult> RehydrateAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            ImmutableHashSet<RenameLocation> locations = null;
+            ImmutableArray<ReferenceLocation> implicitLocations = default;
+            ImmutableArray<ISymbol> referencedSymbols = default;
+
+            if (Locations != null)
+            {
+                using var _ = ArrayBuilder<RenameLocation>.GetInstance(out var builder);
```

Comment:
Is there any way to hint to the ArrayBuilder the size we'll need? Since it's going to be that same size as `Locations` it seems slightly more efficient to allocate that early

Reply:
done!